sos/upload/__init__.py

```diff
@@ -0,0 +1,293 @@
+            if self.opts.upload_profile == "redhat":
+                return self.load_upload_profile(self.profiles_data['redhat'])
+            if self.opts.upload_profile == "ubuntu":
+                return self.load_upload_profile(self.profiles_data['ubuntu'])
+        return self.determine_local_profile()
+
+    def load_upload_profile(self, profile):
+        try:
+            upload_mod = importlib.import_module(
+                profile['mod']
+            )
+            upload_class = getattr(upload_mod, profile['class'])
+            upload = (
+                upload_class(
+                            parser=self.parser,
+                            args=self.args,
+                            cmdline=self.cmdline
+                )
+            )
+            upload.set_commons(self.get_commons())
+            return upload
+        except ImportError as e:
+            self.ui_log.error(f"{profile['description']} upload "
+                              f"profile not available: {e}")
+            return None
+
+    def determine_local_profile(self):
+        self.ui_log.info(
+            "Trying to determine local upload profile..."
+        )
+
+        from sos.policies.distros.redhat import RHELPolicy
+        from sos.policies.distros.ubuntu import UbuntuPolicy
+
+        if isinstance(self.policy, RHELPolicy):
+            return self.load_upload_profile(self.profiles_data['redhat'])
+        if isinstance(self.policy, UbuntuPolicy):
+            return self.load_upload_profile(self.profiles_data['ubuntu'])
+        self.ui_log.info(
+            "Couldn't determine local profile. Using generic profile."
+        )
+        return self.load_upload_profile(self.profiles_data['default'])
+
+    def pre_work(self):
+        # This method will be called before upload begins
+        self.commons = self.get_commons()
+        cmdline_opts = self.commons['cmdlineopts']
+
+        if cmdline_opts.low_priority:
+            self.policy._configure_low_priority()
+
+    def execute(self):
+        self.pre_work()
+        if self.from_cmdline:
+            self.intro()
+            self.archive = self.opts.upload_file
+        self.upload_profile = self.set_upload_profile()
+        try:
+            if os.stat(self.archive).st_size > 0:
+                if os.path.isfile(self.archive):
+                    try:
+                        # Check local profile that was either selected
+                        # via the command line option, or detected locally
+                        if self.upload_profile:
+                            self.upload_profile.upload_archive(self.archive)
+                        else:
+                            # There was no upload profile set, so we'll throw
+                            # an error here
+                            self.ui_log.error(
+                                "No upload profile found. Exiting"
+                            )
+                            sys.exit(1)
```

Comment:
From programming point of view, this "if no profile, then rais error and exit" check shall be executed just after determining the profile (just before the try block). Otherwise we redundantly call the commands in between.

Reply:
Ack, let me upload the fixes based on the way we do things in collect and then I'll revisit this part